app/services/word_watcher.rb

```diff
@@ -54,13 +57,29 @@ def should_flag?
     word_matches_for_action?(:flag)
   end
 
-  def should_block?
-    word_matches_for_action?(:block)
+  def should_block?(all_matches: false)
+    word_matches_for_action?(:block, all_matches: all_matches)
   end
 
-  def word_matches_for_action?(action)
-    r = self.class.word_matcher_regexp(action)
-    r ? r.match(@raw) : false
+  def word_matches_for_action?(action, all_matches: false)
+    regexp = self.class.word_matcher_regexp(action)
+    if regexp
+      match = regexp.match(@raw)
+      return match if !all_matches || !match
+      matches = []
+      regexps = self.class.get_cached_words(action).map do |w|
```

Comment:
This could have a bad performance impact. On a site with 1000 watched words, it looks like each post will loop through each word, make a regexp and run it against the raw?

Is there a reason we can't use the `word_matcher_regexp` to find all the matches?

Reply:
I'm worried about performance too, but `word_matcher_regexp` only finds the first blocked word in raw and not all of the blocked words. For example if raw was `word1 word0` and we had `word0` and `word1` as blocked words, `word_matcher_regexp` would just give us `word1`, whereas we want both `word1` and `word0`.

I tried to come up with a single regexp that combines all blocked words and gives all matches, everything I tried didn't work, and I'm not sure if that's even possible with regular expressions?

Reply:
Can you not call `string.scan(regexp)` to return a list of all the matches in one go?

If not, `match` does take an index of the last match. Your loop would be much more efficient if you only searched from the last match onwards.

Reply:
`String#scan` seems to do exactly what we want, thanks!!!! I'll change the code to use `scan` when I'm back at my computer in a couple of hours.

Reply:
Ok I've update the PR to use `scan`, let me know what you think. (the second commit below has the changes for this)